RFC-0007-plan-constraints-grammar.md

```diff
@@ -0,0 +1,160 @@
+# **RFC007 for Presto**
+
+## Grammar for Plan Constraints
+
+Proposers
+
+* @aaneja
+* @ClarenceThreepwood
+
+## Related Issues
+
+* [Overview doc](https://prestodb.io/wp-content/uploads/Search-Space-Improvements-Plan-Constraints.pdf) on Plan Constraints as a tool to control search space
+* PrestoDB blog on - [Elevating Presto Query Optimization](https://prestodb.io/blog/2024/03/21/elevating-presto-query-optimization/)
+
+## Summary
+
+This document proposes a grammar for specifying plan constraints
+
+## Background
+
+Plan constraints can be used to lock down critical aspects of an execution plan, such as access method, join method, and join order. 
+
+
+## Proposed Implementation
+
+We propose a grammar for specifying independent plan constraints, which take the form of a SQL comment block that would build an object graph of the constraints. Multiple constraints can be specified in a single place. The grammar is open for extension as we develop more mechanisms to lock down plans.
+
+In this first cut, users can build constraints to control 
+- Join orders and distributions for INNER JOIN's
+- Cardinality (row counts) for base relations and join sub-plans
+
+### Grammar (ANTLR 4)
+
+```
+grammar PlanConstraints;
+
+// Lexer rules
+WS         : [ \t\r\n]+ -> skip ;
+NUMBER     : [0-9]+ ;
+LOJ        : 'LOJ' ;
+ROJ        : 'ROJ' ;
+IJ         : 'IJ' ;
+LPAREN     : '(' ;
+RPAREN     : ')' ;
+JOIN_DIST_PARTITIONED : '[P]' ;
+JOIN_DIST_REPLICATED : '[R]' ;
+CARD : 'CARD' ;
+JOIN : 'JOIN' ;
+CONSTRAINTS_START_MARKER :'/*!';
+CONSTRAINTS_END_MARKER :'*/';
+
+fragment DIGIT
+    : [0-9]
+    ;
+
+fragment LETTER
+    : [A-Z]
+    | [a-z]
+    ;
+
+IDENTIFIER
+    : (LETTER | '_') (LETTER | DIGIT | '_' | '@' | ':')*
+    ;
+
+identifier
+    : IDENTIFIER
+    ;
+
+standAloneRelation
+    : identifier
+    ;
+
+groupedRelation
+    : LPAREN joinedRelation RPAREN
+    ;
+
+joinTypeOrDefault
+    : joinType
+    | { "IJ" } // Default value
+    ;
+
+joinType
+    : LOJ
+    | ROJ
+    | IJ
+    ;
+
+joinAttribute
+    : JOIN_DIST_PARTITIONED
+    | JOIN_DIST_REPLICATED
+    ;
+
+joinedRelation
+    : standAloneRelation joinTypeOrDefault standAloneRelation (joinAttribute)? #ss
+    | standAloneRelation joinTypeOrDefault groupedRelation (joinAttribute)? #sg
+    | groupedRelation joinTypeOrDefault standAloneRelation (joinAttribute)? #gs
+    | groupedRelation joinTypeOrDefault groupedRelation (joinAttribute)? #gg
+    ;
+
+
+cardinalityConstraint
+    : CARD LPAREN joinedRelation NUMBER RPAREN
+    | CARD LPAREN standAloneRelation NUMBER RPAREN
+    ;
+
+joinConstraint : JOIN LPAREN joinedRelation RPAREN;
+
+planConstraint
+    : joinConstraint
+    | cardinalityConstraint;
+
+planConstraintString : CONSTRAINTS_START_MARKER (planConstraint)* CONSTRAINTS_END_MARKER;
+
+
+```
+
+
+### Examples of constraints
+
+1. Inner Join constraints - 
```

Comment:
Why can't we just use replicated or paritioned syntax? Imo the brackets and` join ((a c [R]) b [P])` might get complicated quickly

Reply:
Can you elaborate on what this would look like for the cited example ?

Reply:
Like [dremio](https://docs.dremio.com/cloud/reference/sql/commands/SELECT/#distributing-data-evenly-across-execution-engines-during-joins) or [databricks](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-qry-select-hints.html#join-hints).

Current way also is fine but suggestion is to use explicit names for broadcast/partition etc since the goal of this is to allow users to explictly set the types

Reply:
I'm not opposed to this; my only gripe with it is that it that the plan constraint string can get quite verbose, e.g for a 4-table join order - ` join (d ((a c [REPLICATED]) b [PARTITIONED]) [PARTITIONED] )`

Reply:
Do we need this fine grained kind of control.  I like the simplicity of just specifying the partitioning of the table, and every other example I found seems to take that approach. 
1. [Spark](https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-hints.html#join-hints): `BROADCAST(T1)`
2. [Oracle](https://docs.oracle.com/cd/B10500_01/server.920/a96533/hintsref.htm#5544): `PQ_DISTRIBUTE(s BROADCAST, NONE)` (which I also find very clunky).
3. [Vertica](https://www.vertica.com/docs/10.0.x/HTML/Content/Authoring/SQLReferenceManual/LanguageElements/Hints/distrib.htm): you specify inline distrib(L,B)
4. [Dremio](https://docs.dremio.com/cloud/reference/sql/commands/SELECT/#distributing-data-evenly-across-execution-engines-during-joins): mark `BROADCAST` inline

You can specify to use syntactic join order if you want to control the join ordering in a more complex way.

I find the existing syntax complex and hard to use.

Reply:
[DB2](https://www.ibm.com/docs/en/db2/11.5?topic=requests-any-join) supports 'join requests' similar to these join hints. It has variants that specify the join method too
Using a join hint allow us to :
1. Specify partial join order - you don't need to force syntactic join on parts on the rest of the query to force a specifc join shape for part of the query tree
2. Hardcode join orders for tools that generate SQL and cannot guarantee syntactic SQL generation

Reply:
Got it.  I see the appeal of specifying a partial join order for auto generated queries, but also would like it to be easy to specify join hints for the common case where people just want to mark some table for broadcast or similar. I wonder if there's an alternative approach we can use to achieve both of these goals

Reply:
We could add Spark style independent hints -
1. `BROADCAST(T1)` - Always broadcast T1 in any join
2. `BROADCAST(a b)` - If the optimizer chooses a logical join of `a` and b`, and this is a sub-tree of another join graph, use BROADCAST for the join distribution

These will be complementary to the join-order syntax. So the below join hints  are equivalent but provide flexibility to the user -
```
BROADCAST(a b)  join(c (ab))
join (c (a b)[R])
```

Will think of more examples and incorporate them